src/test/scala/tech/cryptonomic/conseil/tezos/TezosDatastoreTest.scala

```diff
@@ -0,0 +1,386 @@
+        //given
+        implicit val randomSeed = RandomSeed(testReferenceTimestamp.getTime)
+        //custom hash generator with predictable seed
+        val generateHash: Int => String = alphaNumericGenerator(new Random(randomSeed.seed))
+
+        val maxLevel = 5
+        val accountsPerBlock = 3
+
+        val generatedBlocks = generateBlocks(maxLevel, testReferenceDateTime)
+
+        val blockAccountsMap = generatedBlocks.map { block =>
+          block -> List.fill(accountsPerBlock)(AccountId(generateHash(5)))
+        }.toMap
+
+        //take the genesis into account
+        val checkpointSize = (maxLevel + 1) * accountsPerBlock
+
+        //set expectations and build the datastore
+        val sut = prepareDatastoreMockWithExpectactions(
+          blocksRepoExpectations = repo => {
+            (repo.writeBlocks _)
+              .expects(*)
+              .returns(dbio(Some(generatedBlocks.size)))
+              .once
+          },
+          operationsRepoExpectations = repo => {
+            (repo.writeOperationsGroups _)
+              .expects(
+                where(
+                  (groups: List[OperationGroupsRow]) => groups.isEmpty
+                )
+              )
+              .returns(dbio(None))
+              .once
+
+            //once only for blocks-related updates
+            (repo.writeUpdates _)
+              .expects(List.empty[BalanceUpdatesRow], None)
+              .returns(dbio(Option.empty[Int]))
+              .once
+
+          },
+          accountsRepoExpectations = repo => {
+            (repo.writeAccountsCheckpoint _)
+              .expects(
+                where(
+                  (accountUpdates: List[(TezosTypes.BlockHash, Int, List[AccountId])]) =>
+                    accountUpdates.size == generatedBlocks.size &&
+                      accountUpdates.forall { case (_, _, ids) => ids.size == accountsPerBlock }
+                )
+              )
+              .returns(dbio(Some(checkpointSize)))
+
+          }
+        )
+
+        //when
+        val action = sut.storeBlocksAndCheckpointAccounts(blockAccountsMap)
+
+        //then
+        testDb.run(action).futureValue shouldBe (Some(generatedBlocks.size), Some(checkpointSize))
+      }
+
+      "store all voting data for multiple blocks" in {
+        //given
+        implicit val randomSeed = RandomSeed(testReferenceTimestamp.getTime)
+
+        val (expectedProposals, expectedRolls, expectedBallots) = (3, 2, 4)
+
+        val generatedBlock = generateSingleBlock(atLevel = 1, atTime = testReferenceDateTime)
+        val proposalToStore = Voting.generateProposals(expectedProposals, forBlock = generatedBlock)
+        val rollsToStore = (generatedBlock, Voting.generateBakersRolls(expectedRolls))
+        val ballotsToStore = (generatedBlock, Voting.generateBallots(expectedBallots))
+
+        //set expectations and build the datastore
+        val sut = prepareDatastoreMockWithExpectactions(
```

Comment:
I'm not a fan of using so many stubs in one test, especially with `expects(*)` which doesn't test the correctness of passing parameters. On the other hand, using the exact matches will tight coupling. 

I'm wondering if using some real implementation (created especially for tests) of those traits would increase the readability of this test. They could be initialized with a more descriptive way.

Reply:
You might be right here, yet I needed to produce a working solution fast, to proceed with work built upon this refactor.
I'm simply using those as stubs to make the test run... I'm not even sure these tests are valuable apart from coverage...

In any case I'm not against make the checks more fine-grained in a separate PR, only I won't put my time to it now, unfortunately.